### DISCUSSION TITLE
[V3] Anyway to control all Livewire routes ?

### DISCUSSION THREAD

**Discussion by felahdab** [livewire/livewire#6771] [1. Help]
Hi.

In livewire V2, the routes used where all registered thru the LivewireServiceProvider.

It had some limitations, but it also had the advantage of having all routes defined in a single place. It was therefore rather easy to change  the URL used by livewire components (for example by making livewire/livewire non discovered, creating your own ServiceProvider extending the LivewireServiceProvider, and in that service provider, change the route registering method to your need).
The config/livewire.php allows to define the app_url parameter which is provided to the underlying alpine.js code to adjust the url used by the javascript parts of the components.

This is a method I am using in a project where multiple apps are hosted in various subpath of a unique domain (a constraint I cannot change). In this project, I have to add this subpath prefix to all routes. 
Everything fits properly, and all packages relying on livewire benefit from this application-wide configuration.

V3 introduces the setUpdateRoute method which is supposed to provide some level of control over the used routes.
Sounds good.

However, the setUpdateRoute method only take care of the livewire.update route but doesn't take care of the upload-file, preview-file or livewire.js routes which are now registered in various Mechanisms...

Hence my question:
- Is there anyway to control all Livewire routes in V3 (for example for adding a prefix for all thoses routes) ?
- And is there a way to remove the default route (the setUpdateRoute method seems to add a new Route but doesn't remove the default one).

Regards.
Florian

**Comment by nerijuso** [comment]
@felahdab were you able to find the solution? As you said setUpdateRoute creates a new route and I am not able to cache routes because there is an old one with the same name.

**Reply by felahdab** [reply]
Hi.

I submitted pr #6803 but it was rejected until I fix additionnal details.
You can however integrate the commit to a project while I try to find the time to do that.

Just yesterday, I resorted to another method which involves a little bit more work but fits my need. 
It is based on the method described in pr #2662 (and the associated article posted by @schakko.

Basically:
- you prevent the livewire service provider to be discovered by Laravel (composer.json)
- you create in your application modified SupportFileUploads, FrontendAssets and HandleRequests. Those must be modified to register the routes as you see fit instead of the basic routes.
- you created your own service provider which extends the base service provider.
- in this service provider, you take the protected function getMechanisms and protected function bootFeatures methods as they are in the base service provider.
- you register this new service provider in your application (config/app.php)

I will post my files here later today.
I haven't tested if route caching works fine with this method (it should), but I don"t have multiple livewire routes anymore at least (and more importantly: my application works and I totally control the URL scheme).

**Reply by felahdab** [reply]
Hello again.

I uploaded the relevant files in the zip file hereunder.
These files should be extracted in the app\Providers folder.
And you would also have to add the dont-discover in composer.json, in the extra property.
`
{
        "laravel": {
            "dont-discover": [
                "livewire/livewire"
            ]
        }
    }`

[PrefixedRoutesLivewire.zip](https://github.com/livewire/livewire/files/13461443/PrefixedRoutesLivewire.zip)

We all this in place, the livewire routes will be prefixed with the config('livewire.route_prefix') value that you should set in the config file.

Hope this helps.

**Reply by felahdab** [reply]
And apparently, route caching works. But you'll have to check. And I would be interested in your feedback.

**Reply by nerijuso** [reply]
I got your idea of how to do a fix. Thanks a lot. I will try your solution.

**Reply by stevebrainng** [reply]
> And apparently, route caching works. But you'll have to check. And I would be interested in your feedback.

Hi @felahdab thanks for your provided work around. I tried it out and encounted error while running composer dump-autoload or even when installing any additional composer package because of the "{ "laravel": { "dont-discover": [ "livewire/livewire" ] } }"

Here's what I'm getting:
![image](https://github.com/livewire/livewire/assets/24863248/3b7ef693-d5a3-427d-8939-a2ef04cbb3bb)

Any suggestions? I'm also interested in the PR you pushed before. 

Regards.

**Reply by felahdab** [reply]
My workaround works with a replacement service provider that you have to register manually in your application.

This service provider extends the normal livewire service provider just to replace the pieces that must bé modified to register the routes as you wish.

The error you get suggests that this replacement service provider is not registered, and consequently the "livewire" binding is undefined.

I think this custom service provider was in the zip file included in the discussions about this PR.

I have been using this workaround since then and composer operations work as usual.

Regards.
